sparse/__init__.py

```diff
@@ -1,323 +1,56 @@
-    "e",
-    "einsum",
-    "elemwise",
-    "empty",
-    "empty_like",
-    "equal",
-    "exp",
-    "expand_dims",
-    "expm1",
-    "eye",
-    "finfo",
-    "flip",
-    "float16",
-    "float32",
-    "float64",
-    "floor",
-    "floor_divide",
-    "full",
-    "full_like",
-    "greater",
-    "greater_equal",
-    "iinfo",
-    "imag",
-    "inf",
-    "int16",
-    "int32",
-    "int64",
-    "int8",
-    "isfinite",
-    "isinf",
-    "isnan",
-    "isneginf",
-    "isposinf",
-    "kron",
-    "less",
-    "less_equal",
-    "load_npz",
-    "log",
-    "log10",
-    "log1p",
-    "log2",
-    "logaddexp",
-    "logical_and",
-    "logical_not",
-    "logical_or",
-    "logical_xor",
-    "matmul",
-    "matrix_transpose",
-    "max",
-    "mean",
-    "min",
-    "moveaxis",
-    "multiply",
-    "nan",
-    "nanmax",
-    "nanmean",
-    "nanmin",
-    "nanprod",
-    "nanreduce",
-    "nansum",
-    "negative",
-    "newaxis",
-    "nonzero",
-    "not_equal",
-    "ones",
-    "ones_like",
-    "outer",
-    "pad",
-    "permute_dims",
-    "pi",
-    "positive",
-    "pow",
-    "prod",
-    "random",
-    "real",
-    "remainder",
-    "reshape",
-    "result_type",
-    "roll",
-    "round",
-    "save_npz",
-    "sign",
-    "sin",
-    "sinh",
-    "sort",
-    "sqrt",
-    "square",
-    "squeeze",
-    "stack",
-    "std",
-    "subtract",
-    "sum",
-    "take",
-    "tan",
-    "tanh",
-    "tensordot",
-    "tril",
-    "triu",
-    "trunc",
-    "uint16",
-    "uint32",
-    "uint64",
-    "uint8",
-    "unique_counts",
-    "unique_values",
-    "var",
-    "vecdot",
-    "where",
-    "zeros",
-    "zeros_like",
-]
 
-__array_api_version__ = "2022.12"
+class BackendType(Enum):
+    pydata = "pydata"
+    finch = "finch"
+
+
+_ENV_VAR_NAME = "SPARSE_BACKEND"
+
+backend_var = ContextVar("backend", default=BackendType.pydata)
+
+if _ENV_VAR_NAME in os.environ:
+    backend_var.set(BackendType[os.environ[_ENV_VAR_NAME]])
+
+
+class Backend:
+    def __init__(self, backend=BackendType.pydata):
+        self.backend = backend
+        self.token = None
+
+    def __enter__(self):
+        token = backend_var.set(self.backend)
+        self.token = token
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        backend_var.reset(self.token)
+        self.token = None
+
+    @staticmethod
+    def get_backend_module():
+        backend = backend_var.get()
+        if backend == BackendType.pydata:
+            import sparse.pydata_backend as backend_module
+        elif backend == BackendType.finch:
+            import sparse.finch_backend as backend_module
+        else:
+            raise ValueError(f"Invalid backend identifier: {backend}")
+        return backend_module
+
+
+def __getattr__(attr):
+    if attr == "pydata_backend":
+        import sparse.pydata_backend as backend_module
+
+        return backend_module
+    if attr == "finch_backend":
+        import sparse.finch_backend as backend_module
```

Comment:
Shouldn't these refer to `backend_var`?

Reply:
Ah, I see what you're doing. I'd just import these at the top so they work out of the box:

```python
from . import pydata_backend, finch_backend
```

Reply:
So the reason we have these lines is a special case when a user does e.g. `sparse.pydata_backend`. For anything else, e.g. `sparse.permute_dims` the `return getattr(...` line is used.

I can't import them at the top because these backends must be imported lazily. Importing `finch_backend` when Finch.jl is not installed will fail.

Reply:
Ah, that's true.